src/Processors/Transforms/Streaming/JoinTransformWithAlignment.h

```diff
@@ -27,81 +30,120 @@ class JoinTransformWithAlignment final : public IProcessor
     Status prepare() override;
     void work() override;
 
+    void checkpoint(CheckpointContextPtr ckpt_ctx) override;
+    void recover(CheckpointContextPtr ckpt_ctx) override;
+
     static Block transformHeader(Block header, const HashJoinPtr & join);
 
 private:
-    Status prepareLeftInput();
-    Status prepareRightInput();
-
-    inline Int64 getRightWatermark(const Chunk & chunk) const
+    struct InputPortWithData
     {
-        return getWatermark(chunk, right_watermark_column_position, right_watermark_column_type);
-    }
+        explicit InputPortWithData(InputPort * input_port_) : input_port(input_port_) { }
 
-    inline Int64 getLeftWatermark(const Chunk & chunk) const
-    {
-        return getWatermark(chunk, left_watermark_column_position, left_watermark_column_type);
-    }
+        InputPort * input_port;
 
-    inline Int64 getWatermark(const Chunk & chunk, std::optional<size_t> col_pos, const DataTypePtr & col_type) const
-    {
-        if (col_pos)
+        /// Input state
+        /// NOTE: Assume the input chunk is time-ordered
+        /// pair-<retracted_chunk, chunk>
+        SERDE std::list<std::pair<LightChunk, LightChunkWithTimestamp>> input_chunks;
```

Comment:
We don't need the pair ? It seems unnecessary introduce complexity and inefficiency. Just std::list<LightChunkWithTimestamp> will do the work ? If first chunk is retracted, just append it and then wait for the following append chunk.

Reply:
The LightChunkWithTimestamp don't contain chunk_ctx, if so we need extra flag to mark it

Reply:
and for retracted data, it's min/max timestamp is also late and bad.

Reply:
LightChunkWithTimestamp is a super struct of LightChunk. How comes it wont work.